### DISCUSSION TITLE
Enable modification of life_stage variable for other organisms

### DISCUSSION THREAD

**Discussion by habacucfm** [traitecoevo/traits.build#157] [Ideas]
Hello,
I would like to request an enhancement to the traits.build package to allow the life_stage variable to be modifiable for other organisms. Currently, it seems the variable is restricted to plants, and expanding its flexibility would be beneficial for broader ecological and evolutionary studies.

Proposed Solution:
Allow users to define and modify the life_stage variable to include a variety of life stages pertinent to different organisms, beyond the current constraints.

This change will make the traits.build package more versatile and applicable to a wider range of ecological and evolutionary research.
Researchers working with diverse organisms will be able to accurately represent and analyze life stage data.

Thank you for considering this enhancement.

**Comment by dfalster** [comment]
(NB: I've converted this issue into a Discussion, as we'll be using Discussions more going forward). 

Hi @habacucfm !

Thanks for your interest and suggestion.  Yes, we certainly want the system to be applicable to other taxa. My understanding is that you can already achieve what you want. We have a `life_stage` variable, but I think you can use whatever terms you like. Moreover, this term is meant to be compatible with the Darwin core equivalent (https://dwc.tdwg.org/terms/#dwc:lifeStage). 

We'll need to confirm this and that will take some time, as @ehwenk is on leave throughout July.

In the meantime, can you point me to whatever material gave you the impression this was restricted to plants?

**Comment by habacucfm** [comment]
Thank you @dfalster! I wrongfully assumed that since `life_stage` did not seem to appear in the traits.build template `config/traits.yml`
and was defined here https://github.com/traitecoevo/traits.build/blob/f2e01eaedbb464e0e795fd394c0c731514747f2a/inst/support/traits.build_schema.yml#L70   it would be locked in on plants. 

I ran a test and life_stage runs with different definitions as you suggest. Thank you!

However, I noticed that `data_test("study")`  does not flag differences in life_stage controlled vocabularies between raw data file and config/traits.yml

Example below:

```
# clone https://github.com/traitecoevo/traits.build-template.git
library(traits.build) #load 

#load example 1 an add mock life_stage column called example_Column with random categories

categories <- c("egg", "larva", "C", "D", "E")
dat <- read_csv("data/example_dataset_1/data.csv") %>%
  mutate(example_Column = sample(categories, size = n(), replace = TRUE))
write_csv(x = dat, "data/example_dataset_1/data.csv")
#Manually modify metadata.yml file in  "data/example_dataset_1" add example_Column as life_stage 

#Manually modify conflict/traits.yml file adding a different controlled vocabulary for lifestage
# life_stage:
#   label: Insect life stage
# description: The developmental stage of an insect insect.
# comments: .na
# type: categorical
# allowed_values_levels:
#   egg: Initial stage where the insect begins its life
# larva: The second stage (e.g. caterpillars, maggots, grubs)
# pupa: Third stage where the insect undergoes transformation
# nymph: Second stage for hemimetabolic insects, nymphs resemble smaller versions of the adults
# adult: The last stage where the insect is capable of reproduction
# entity_URI: .na

#test dataset
dataset_test("example_dataset_1")
#Throws unrelated error about context being out of bounds
# [ FAIL 1 | WARN 0 | SKIP 0 | PASS 49 ]
# 
# ── Error (???): example_dataset_1 ──────────────────────────────────────────────
# <subscriptOutOfBoundsError/error/condition>
#   Error in `FUN(X[[i]], ...)`: subscript out of bounds
#Build pipeline
build_setup_pipeline(method = "base", database_name = "traits.build_database")
#source file
source("build.R")

#Check example 1 life stage column
example_dataset_1$traits  %>%
  distinct(life_stage) 
#Names have been propagated from raw data but they do not match controlled vocabulary in config/traits.yml ```

**Reply by dfalster** [reply]
Ok good, I'll ask @ehwenk to follow up on these issues when she's back from travel.

**Comment by ehwenk** [comment]
@habacucfm Thanks for your question about traits.build. 

As @dfalster said, the field life_stage aligns with the Darwin Core https://dwc.tdwg.org/terms/#dwc:lifeStage. As with Darwin Core we have, at this point, chosen not to use a controlled vocabulary explicitly because such disparate terms are used for different taxonomic groups. For fields where a controlled vocabulary is used, the allowed values are listed at the top of the traits.build schema file (https://github.com/traitecoevo/traits.build/blob/develop/inst/support/traits.build_schema.yml) - for instance for `entity_type`, `value_type`, `basis_of_record`, and `basis_of_value`. The same schema file, in theory, underpins all traits.build databases.

The file https://github.com/traitecoevo/austraits.build/blob/develop/config/traits.yml is the trait dictionary for a specific database and includes trait definitions and trait values that apply to a database. As `life_stage` indicates metadata that accompanies a trait measurement, but isn't a trait itself, it isn't included in a databases' trait dictionary. What constitutes a "trait" is very much up to an individual database curator, and you can of course define `life_stage_trait` (or any other name that isn't `life_stage`) as a trait, with set values, but then you would be entering life stage data as a "trait" not within the `life_stage` field. You should currently be able to have a column of different life stage values that you specify should be read in through a dataset's `metadata.yml` file. I'm happy to help work through some examples of this.
